cocotb/share/lib/vhpi/VhpiCbHdl.cpp

```diff
@@ -916,81 +916,64 @@ VhpiNextPhaseCbHdl::VhpiNextPhaseCbHdl(GpiImplInterface *impl) : GpiCbHdl(impl),
     cb_data.time = &vhpi_time;
 }
 
-void vhpi_mappings(GpiIteratorMapping<vhpiClassKindT, vhpiOneToManyT> &map)
-{
-    /* vhpiRootInstK */
-    vhpiOneToManyT root_options[] = {
-        vhpiInternalRegions,
-        vhpiSigDecls,
-        vhpiVarDecls,
-        vhpiPortDecls,
-        vhpiGenericDecls,
-        vhpiConstDecls,
-        //    vhpiIndexedNames,
-        vhpiCompInstStmts,
-        vhpiBlockStmts,
-        (vhpiOneToManyT)0,
-    };
-    map.add_to_options(vhpiRootInstK, &root_options[0]);
-
-    /* vhpiSigDeclK */
-    vhpiOneToManyT sig_options[] = {
-        vhpiIndexedNames,
-        vhpiSelectedNames,
-        (vhpiOneToManyT)0,
-    };
-    map.add_to_options(vhpiGenericDeclK, &sig_options[0]);
-    map.add_to_options(vhpiSigDeclK, &sig_options[0]);
-
-    /* vhpiIndexedNameK */
-    map.add_to_options(vhpiSelectedNameK, &sig_options[0]);
-    map.add_to_options(vhpiIndexedNameK, &sig_options[0]);
-
-    /* vhpiCompInstStmtK */
-    map.add_to_options(vhpiCompInstStmtK, &root_options[0]);
-
-    /* vhpiSimpleSigAssignStmtK */
-    vhpiOneToManyT simplesig_options[] = {
-        vhpiDecls,
-        vhpiInternalRegions,
-        vhpiSensitivitys,
-        vhpiStmts,
-        (vhpiOneToManyT)0,
-    };
-    map.add_to_options(vhpiCondSigAssignStmtK, &simplesig_options[0]);
-    map.add_to_options(vhpiSimpleSigAssignStmtK, &simplesig_options[0]);
-    map.add_to_options(vhpiSelectSigAssignStmtK, &simplesig_options[0]);
-
-    /* vhpiPortDeclK */
-    map.add_to_options(vhpiPortDeclK, &sig_options[0]);
-
-    /* vhpiForGenerateK */
-    vhpiOneToManyT gen_options[] = {
-        vhpiDecls,
-        vhpiInternalRegions,
-        vhpiSigDecls,
-        vhpiVarDecls,
-        vhpiConstDecls,
-        vhpiCompInstStmts,
-        vhpiBlockStmts,
-        (vhpiOneToManyT)0,
+decltype(VhpiIterator::iterate_over) VhpiIterator::iterate_over = []{
+    /* for reused lists */
+    std::initializer_list<vhpiOneToManyT> root_options;
+    std::initializer_list<vhpiOneToManyT> sig_options;
+    std::initializer_list<vhpiOneToManyT> simplesig_options;
+    std::initializer_list<vhpiOneToManyT> gen_options;
+
+    return decltype(VhpiIterator::iterate_over) {
+        {vhpiRootInstK, root_options = {
+            vhpiInternalRegions,
+            vhpiSigDecls,
+            vhpiVarDecls,
+            vhpiPortDecls,
+            vhpiGenericDecls,
+            vhpiConstDecls,
+            //    vhpiIndexedNames,
+            vhpiCompInstStmts,
+            vhpiBlockStmts,
+        }},
+        {vhpiCompInstStmtK, root_options},
+
+        {vhpiGenericDeclK, sig_options = {
+            vhpiIndexedNames,
+            vhpiSelectedNames,
+        }},
+        {vhpiSigDeclK, sig_options},
+        {vhpiSelectedNameK, sig_options},
+        {vhpiIndexedNameK, sig_options},
+        {vhpiPortDeclK, sig_options},
```

Comment:
This assign and reuse trick feels a little sneaky, but it's a nice way to keep the initialization and use together.